File: java/src/main/java/io/valkey/javabenchmark/engine/BenchmarkEngine.java
/*
 * Copyright 2025 the original author or authors.
 */
package io.valkey.javabenchmark.engine;

import io.valkey.javabenchmark.client.BenchmarkClient;
import io.valkey.javabenchmark.client.BenchmarkClientFactory;
import io.valkey.javabenchmark.command.Command;
import io.valkey.javabenchmark.command.Command.CommandResult;
import io.valkey.javabenchmark.command.CommandFactory;
import io.valkey.javabenchmark.config.*;
import io.valkey.javabenchmark.metrics.NdjsonMetricsWriter;
import io.valkey.javabenchmark.metrics.MetricsCollector;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Main benchmark engine that orchestrates the benchmark execution.
 * 
 * <p>Uses a <b>virtual-thread-per-client</b> architecture where each client connection
 * is owned by a long-lived virtual thread that runs a tight request loop. This design
 * eliminates the overhead of per-request VT creation and scheduling that limited the
 * previous architecture to ~400K RPS.</p>
 * 
 * <h3>Architecture</h3>
 * <pre>
 *   Main thread
 *     │
 *     ├── Creates N client connections
 *     ├── Spawns N long-lived virtual threads (one per client)
 *     │     │
 *     │     ├── VT-0: while(running) { execute → join → record → repeat }
 *     │     ├── VT-1: while(running) { execute → join → record → repeat }
 *     │     ├── ...
 *     │     └── VT-N: while(running) { execute → join → record → repeat }
 *     │
 *     ├── Logs progress periodically
 *     └── Waits for all VTs to complete, merges metrics
 * </pre>
 * 
 * <p>Key advantages over the previous command-issuer + semaphore design:</p>
 * <ul>
 *   <li><b>No per-request VT creation</b> — eliminates 400K+ VT spawns/sec overhead</li>
 *   <li><b>No semaphore contention</b> — backpressure is implicit (VT blocks on I/O)</li>
 *   <li><b>No CompletableFuture chains</b> — direct join() parks VT, not carrier</li>
 *   <li><b>Fewer carrier threads needed</b> — VTs park on I/O, carriers are free</li>
 *   <li><b>Per-VT state</b> — KeyGenerator, CommandSelector are thread-local, zero contention</li>
 * </ul>
 *
 * @author Ilia Kolominsky
 */
public class BenchmarkEngine {
    private static final Logger logger = LoggerFactory.getLogger(BenchmarkEngine.class);
    
    /** Default pipeline depth (max in-flight requests per client) */
    private static final int DEFAULT_PIPELINE_DEPTH = 1;
    
    /** Progress logging interval in seconds */
    private static final int PROGRESS_LOG_INTERVAL_SECONDS = 10;

    private final String host;
    private final int port;
    private final DriverConfig driverConfig;
    private final WorkloadConfig workloadConfig;
    private final NdjsonMetricsWriter metricsWriter;
    private final String commitId;
    
    // Client reference for version info
    private BenchmarkClient sampleClient;

    public BenchmarkEngine(String host, int port, DriverConfig driverConfig,
                          WorkloadConfig workloadConfig, String metricsPath) {
        this(host, port, driverConfig, workloadConfig, metricsPath, null, null);
    }

    public BenchmarkEngine(String host, int port, DriverConfig driverConfig,
                          WorkloadConfig workloadConfig, String metricsPath, String commitId) {
        this(host, port, driverConfig, workloadConfig, metricsPath, commitId, null);
    }

    public BenchmarkEngine(String host, int port, DriverConfig driverConfig,
                          WorkloadConfig workloadConfig, String metricsPath, String commitId,
                          Integer commandIssuerThreads) {
        this.host = host;
        this.port = port;
        this.driverConfig = driverConfig;
        this.workloadConfig = workloadConfig;
        this.metricsWriter = new NdjsonMetricsWriter(metricsPath);
        this.commitId = commitId;
    }

    public void run() throws Exception {
        logger.info("Starting benchmark: {}", workloadConfig.getBenchmarkProfile().getName());
        logger.info("Driver: {}, Mode: {}", driverConfig.getDriverId(), driverConfig.getMode());
        logger.info("Server: {}:{}", host, port);

        // Set up metadata using a sample client to get version info
        setupMetadata();

        for (PhaseConfig phase : workloadConfig.getPhases()) {
            executePhase(phase);
        }

        logger.info("Benchmark completed");
    }

    /**
     * Set up metadata for metrics output, including driver versions.
     */
    private void setupMetadata() {
        try {
            sampleClient = BenchmarkClientFactory.createAndConnect(host, port, driverConfig);
            
            String driverId = driverConfig.getDriverId();
            String primaryVersion = sampleClient.getDriverVersion();
            String secondaryDriverId = driverConfig.getSecondaryDriverId();
            String secondaryVersion = sampleClient.getSecondaryDriverVersion();
            
            metricsWriter.setMetadata(commitId, driverId, primaryVersion, secondaryDriverId, secondaryVersion);
            
            logger.info("Metadata: commit={}, driver={}, version={}", 
                        commitId != null ? commitId : "N/A", driverId, primaryVersion);
            
            sampleClient.close();
            sampleClient = null;
        } catch (Exception e) {
            logger.warn("Failed to get driver version for metadata: {}", e.getMessage());
            metricsWriter.setMetadata(commitId, driverConfig.getDriverId(), "unknown", 
                                     driverConfig.getSecondaryDriverId(), null);
        }
    }

    private void executePhase(PhaseConfig phase) throws Exception {
        logger.info("=== Starting phase: {} ({}) ===", phase.getId(), phase.getDescription());

        int pipelineDepth = phase.getPipelineDepth() > 0 ? phase.getPipelineDepth() : DEFAULT_PIPELINE_DEPTH;
        
        // Create client connections
        List<BenchmarkClient> clients = createClients(phase);
        
        // Create commands
        List<Command> commands = CommandFactory.createAll(phase.getCommands());
        
        // Create key generator (base — each VT will fork its own)
        KeyGenerator keyGenerator = KeyGenerator.create(phase.getKeyspace());
        
        // Create rate limiter
        RateLimiter rateLimiter = phase.hasRpsLimit() ? 
                RateLimiter.create(phase.getRpsLimit()) : null;
        
        // Create metrics collector
        MetricsCollector metrics = new MetricsCollector();

        // Warmup: send PINGs on all clients (blocking, before measured workload)
        warmupClients(clients, phase.getWarmupRequests());

        // Execute workload with VT-per-client architecture
        String status = executeWorkload(phase, clients, commands, keyGenerator, 
                                         rateLimiter, metrics, pipelineDepth);

        // Write metrics
        metricsWriter.writePhaseResults(phase.getId(), status, phase.getConnections(), metrics);

        // Close clients
        closeClients(clients);

        logger.info("=== Phase {} completed: {} ===", phase.getId(), status);
    }

    private List<BenchmarkClient> createClients(PhaseConfig phase) throws Exception {
        logger.info("Creating {} connections...", phase.getConnections());
        
        List<BenchmarkClient> clients = new ArrayList<>();
        RateLimiter cpsLimiter = phase.hasCpsLimit() ? 
                RateLimiter.create(phase.getCpsLimit()) : null;

        for (int i = 0; i < phase.getConnections(); i++) {
            if (cpsLimiter != null) {
                cpsLimiter.acquire();
            }
            
            BenchmarkClient client = BenchmarkClientFactory.createAndConnect(host, port, driverConfig);
            clients.add(client);
            
            if ((i + 1) % 50 == 0) {
                logger.info("Created {}/{} connections", i + 1, phase.getConnections());
            }
        }
        
        logger.info("All {} connections established", clients.size());
        return clients;
    }

    /**
     * Warmup: send PINGs on all clients to verify connections are alive.
     * 
     * @throws Exception if any warmup PING fails (phase should be aborted)
     */
    private void warmupClients(List<BenchmarkClient> clients, int warmupRequestsPerClient) 
            throws Exception {
        if (warmupRequestsPerClient <= 0) return;
        
        logger.info("Warming up {} clients with {} PING(s) each...", 
                    clients.size(), warmupRequestsPerClient);
        
        // Enable warmup mode so clients (e.g., RecordingBenchmarkClient) suppress error simulation
        for (BenchmarkClient client : clients) {
            client.setWarmupMode(true);
        }
        
        try {
            // Use VTs for parallel warmup
            try (var executor = Executors.newVirtualThreadPerTaskExecutor()) {
                List<Future<?>> futures = new ArrayList<>();
                for (BenchmarkClient client : clients) {
                    futures.add(executor.submit(() -> {
                        for (int r = 0; r < warmupRequestsPerClient; r++) {
                            client.ping().join(); // Let exceptions propagate
                        }
                    }));
                }
                // Wait for all warmup tasks — propagate any failure
                for (Future<?> f : futures) {
                    f.get(30, TimeUnit.SECONDS);
                }
            }
        } finally {
            // Disable warmup mode — error simulation resumes for the actual workload
            for (BenchmarkClient client : clients) {
                client.setWarmupMode(false);
            }
        }
        
        logger.info("Warmup completed");
    }

    private String executeWorkload(PhaseConfig phase, List<BenchmarkClient> clients,
                                   List<Command> commands, KeyGenerator keyGenerator,
                                   RateLimiter rateLimiter, MetricsCollector metrics,
                                   int pipelineDepth) {
        
        CompletionConfig completion = phase.getCompletion();
        AtomicBoolean running = new AtomicBoolean(true);
        AtomicLong requestCount = new AtomicLong(0);
        
        int workerCount = clients.size();
        logger.info("Spawning {} worker virtual threads (pipeline_depth={})", 
                    workerCount, pipelineDepth);
        
        // Compute per-worker request budget for request-based completion
        long totalTargetRequests = completion.isRequestBased() ? completion.getTotalRequests() : -1;
        
        // Duration end time
        long endTimeMs = completion.isDurationBased() 
                ? System.currentTimeMillis() + (completion.getDurationSeconds() * 1000)
                : Long.MAX_VALUE;
        
        metrics.start();
        long startTime = System.currentTimeMillis();
        
        try (var executor = Executors.newVirtualThreadPerTaskExecutor()) {
            List<Future<?>> workerFutures = new ArrayList<>(workerCount);
            
            for (int i = 0; i < workerCount; i++) {
                final int workerIndex = i;
                final BenchmarkClient client = clients.get(i);
                final KeyGenerator workerKeyGen = keyGenerator.forkForThread(workerIndex);
                final CommandSelector workerSelector = new CommandSelector(commands);
                
                workerFutures.add(executor.submit(() -> {
                    Thread.currentThread().setName("worker-" + workerIndex);
                    
                    try {
                        if (pipelineDepth <= 1) {
                            runSyncLoop(client, workerSelector, workerKeyGen, rateLimiter,
                                       metrics, requestCount, running, totalTargetRequests, endTimeMs);
                        } else {
                            runPipelinedLoop(client, workerSelector, workerKeyGen, rateLimiter,
                                            metrics, requestCount, running, totalTargetRequests, 
                                            endTimeMs, pipelineDepth);
                        }
                    } catch (Exception e) {
                        // Signal all workers to stop — fail-fast on unexpected errors
                        running.set(false);
                        if (e instanceof RuntimeException re) throw re;
                        throw new RuntimeException(e); // Wrap checked exceptions for Future.get()
                    }
                }));
            }
            
            // Progress logging from main thread while workers run
            long lastLogTime = System.currentTimeMillis();
            boolean allDone = false;
            
            while (!allDone) {
                allDone = true;
                for (Future<?> f : workerFutures) {
                    if (!f.isDone()) {
                        allDone = false;
                        break;
                    }
                }
                if (!allDone) {
                    Thread.sleep(100);
                    lastLogTime = logProgressIfNeeded(requestCount.get(), totalTargetRequests, 
                                                      lastLogTime, startTime);
                }
            }
            
            // Check for worker failures — any exception means the phase failed
            boolean hasWorkerFailures = false;
            for (Future<?> f : workerFutures) {
                try {
                    f.get();
                } catch (ExecutionException e) {
                    logger.error("Worker failed: {}", e.getCause().getMessage());
                    hasWorkerFailures = true;
                }
            }
            
            if (hasWorkerFailures) {
                logger.error("Phase failed: one or more workers crashed");
                return "ERROR";
            }
            
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            logger.warn("Workload interrupted");
            return "INTERRUPTED";
        } catch (Exception e) {
            logger.error("Error during workload execution: {}", e.getMessage());
            return "ERROR";
        } finally {
            metrics.stop();
        }
        
        logger.info("All operations completed. Total requests: {}", requestCount.get());
        return "COMPLETED";
    }

    /**
     * Synchronous request loop for pipeline_depth=1.
     * 
     * <p>Each VT owns one client and runs: execute → join → record → repeat.
     * The join() parks the VT (not the carrier), so the carrier is free to
     * run other VTs while waiting for the server response.</p>
     */
    private void runSyncLoop(BenchmarkClient client, CommandSelector selector,
                              KeyGenerator keyGen, RateLimiter rateLimiter,
                              MetricsCollector metrics, AtomicLong requestCount,
                              AtomicBoolean running, long totalTargetRequests, long endTimeMs) 
            throws InterruptedException {
        
        while (running.get()) {
            // Check termination conditions
            if (System.currentTimeMillis() >= endTimeMs) break;
            if (totalTargetRequests > 0) {
                long claimed = requestCount.getAndIncrement();
                if (claimed >= totalTargetRequests) {
                    requestCount.decrementAndGet();
                    break;
                }
            }
            
            // Rate limiting (parks VT, not carrier)
            if (rateLimiter != null) {
                rateLimiter.acquire();
            }
            
            // Select and execute command
            Command command = selector.select();
            try {
                CommandResult result = command.execute(client, keyGen).join();
                metrics.record(result);
            } catch (CompletionException e) {
                // Command execution failed
                metrics.record(CommandResult.failure(command.getName(), e.getMessage()));
            }
            
            // For duration-based: count here (after execution)
            if (totalTargetRequests <= 0) {
                requestCount.incrementAndGet();
            }
        }
    }

    /**
     * Pipelined request loop for pipeline_depth > 1.
     * 
     * <p>Each VT owns one client and maintains up to pipeline_depth in-flight requests.
     * Uses CompletableFuture.anyOf() to wait for the next completion, then immediately
     * issues a new request to keep the pipeline full.</p>
     */
    private void runPipelinedLoop(BenchmarkClient client, CommandSelector selector,
                                   KeyGenerator keyGen, RateLimiter rateLimiter,
                                   MetricsCollector metrics, AtomicLong requestCount,
                                   AtomicBoolean running, long totalTargetRequests,
                                   long endTimeMs, int pipelineDepth)
            throws InterruptedException {
        
        // Track pending futures with their command names for error reporting
        List<CompletableFuture<CommandResult>> pending = new ArrayList<>(pipelineDepth);
        
        // Fill the pipeline initially
        for (int i = 0; i < pipelineDepth && running.get(); i++) {
            if (!claimAndSubmit(client, selector, keyGen, rateLimiter, requestCount,
                               totalTargetRequests, endTimeMs, pending)) {
                break;
            }
        }
        
        // Event loop: wait for completion, record, refill
        while (!pending.isEmpty() && running.get()) {
            // Wait for any future to complete
            try {
                CompletableFuture.anyOf(pending.toArray(new CompletableFuture[0])).join();
            } catch (CompletionException e) {
                // At least one failed — we'll handle it below
            }
            
            // Process all completed futures
            Iterator<CompletableFuture<CommandResult>> it = pending.iterator();
            while (it.hasNext()) {
                CompletableFuture<CommandResult> f = it.next();
                if (f.isDone()) {
                    it.remove();
                    try {
                        CommandResult result = f.join();
                        metrics.record(result);
                    } catch (CompletionException e) {
                        metrics.record(CommandResult.failure("UNKNOWN", e.getMessage()));
                    }
                    
                    // For duration-based: count completions
                    if (totalTargetRequests <= 0) {
                        requestCount.incrementAndGet();
                    }
                    
                    // Refill the pipeline slot
                    if (running.get()) {
                        claimAndSubmit(client, selector, keyGen, rateLimiter, requestCount,
                                      totalTargetRequests, endTimeMs, pending);
                    }
                }
            }
        }
    }

    /**
     * Claim a request slot (for request-based) and submit a new request.
     * Returns false if the phase should end.
     */
    private boolean claimAndSubmit(BenchmarkClient client, CommandSelector selector,
                                    KeyGenerator keyGen, RateLimiter rateLimiter,
                                    AtomicLong requestCount, long totalTargetRequests,
                                    long endTimeMs, List<CompletableFuture<CommandResult>> pending)
            throws InterruptedException {
        
        if (System.currentTimeMillis() >= endTimeMs) return false;
        
        if (totalTargetRequests > 0) {
            long claimed = requestCount.getAndIncrement();
            if (claimed >= totalTargetRequests) {
                requestCount.decrementAndGet();
                return false;
            }
        }
        
        if (rateLimiter != null) {
            rateLimiter.acquire();
        }
        
        Command command = selector.select();
        pending.add(command.execute(client, keyGen));
        return true;
    }

    private void closeClients(List<BenchmarkClient> clients) {
        logger.info("Closing {} connections...", clients.size());
        for (BenchmarkClient client : clients) {
            try {
                client.close();
            } catch (Exception e) {
                logger.warn("Error closing client: {}", e.getMessage());
            }
        }
    }
    
    /**
     * Log progress if enough time has elapsed since last log.
     */
    private long logProgressIfNeeded(long current, long target, long lastLogTime, long startTime) {
        long now = System.currentTimeMillis();
        if (now - lastLogTime < PROGRESS_LOG_INTERVAL_SECONDS * 1000) {
            return lastLogTime;
        }
        
        long elapsedMs = now - startTime;
        long rate = elapsedMs > 0 ? (current * 1000 / elapsedMs) : 0;
        
        if (target > 0) {
            double percent = (current * 100.0 / target);
            logger.info("Progress: {}/{} requests ({} %) - {} req/s", 
                       current, target, String.format("%.1f", percent), rate);
        } else {
            logger.info("Progress: {} requests - {} req/s", current, rate);
        }
        
        return now;
    }

    /**
     * Selects commands based on their weights.
     */
    private static class CommandSelector {
        private final List<Command> commands;
        private final double[] cumulativeWeights;
        private final Random random = new Random();

        CommandSelector(List<Command> commands) {
            this.commands = commands;
            this.cumulativeWeights = new double[commands.size()];
            
            double sum = 0;
            for (int i = 0; i < commands.size(); i++) {
                sum += commands.get(i).getWeight();
                cumulativeWeights[i] = sum;
            }
        }

        Command select() {
            double r = random.nextDouble();
            for (int i = 0; i < cumulativeWeights.length; i++) {
                if (r <= cumulativeWeights[i]) {
                    return commands.get(i);
                }
            }
            return commands.get(commands.size() - 1);
        }
    }
}

File: csharp/src/RespBench/Engine/BenchmarkEngine.cs
/*
 * Copyright 2025 the original author or authors.
 */
using Microsoft.Extensions.Logging;
using RespBench.Client;
using RespBench.Command;
using RespBench.Config;
using RespBench.Metrics;

namespace RespBench.Engine;

/// <summary>
/// Main benchmark engine that orchestrates the benchmark execution.
/// Uses a Task-per-client architecture where each client connection
/// is owned by a long-lived Task that runs a tight request loop.
/// </summary>
public class BenchmarkEngine
{
    private const int DefaultPipelineDepth = 1;
    private const int ProgressLogIntervalSeconds = 10;

    private readonly string _host;
    private readonly int _port;
    private readonly DriverConfig _driverConfig;
    private readonly WorkloadConfig _workloadConfig;
    private readonly NdjsonMetricsWriter _metricsWriter;
    private readonly string? _commitId;
    private readonly ILogger<BenchmarkEngine> _logger;

    public BenchmarkEngine(string host, int port, DriverConfig driverConfig,
                          WorkloadConfig workloadConfig, string metricsPath,
                          string? commitId = null, int? commandIssuerThreads = null)
    {
        _host = host;
        _port = port;
        _driverConfig = driverConfig;
        _workloadConfig = workloadConfig;
        _metricsWriter = new NdjsonMetricsWriter(metricsPath);
        _commitId = commitId;

        using var loggerFactory = LoggerFactory.Create(b => b.AddConsole().SetMinimumLevel(LogLevel.Information));
        _logger = loggerFactory.CreateLogger<BenchmarkEngine>();
    }

    public async Task Run()
    {
        Console.WriteLine($"Starting benchmark: {_workloadConfig.BenchmarkProfileData?.Name}");
        Console.WriteLine($"Driver: {_driverConfig.DriverId}, Mode: {_driverConfig.Mode}");
        Console.WriteLine($"Server: {_host}:{_port}");

        SetupMetadata();

        foreach (var phase in _workloadConfig.Phases)
        {
            await ExecutePhase(phase).ConfigureAwait(false);
        }

        Console.WriteLine("Benchmark completed");
    }

    private void SetupMetadata()
    {
        try
        {
            using var sampleClient = BenchmarkClientFactory.CreateAndConnect(_host, _port, _driverConfig);
            string driverId = _driverConfig.DriverId;
            string primaryVersion = sampleClient.DriverVersion;
            string? secondaryDriverId = _driverConfig.SecondaryDriverId;
            string? secondaryVersion = sampleClient.SecondaryDriverVersion;

            _metricsWriter.SetMetadata(_commitId, driverId, primaryVersion, secondaryDriverId, secondaryVersion);
            Console.WriteLine($"Metadata: commit={_commitId ?? "N/A"}, driver={driverId}, version={primaryVersion}");
        }
        catch (Exception e)
        {
            Console.WriteLine($"Warning: Failed to get driver version for metadata: {e.Message}");
            _metricsWriter.SetMetadata(_commitId, _driverConfig.DriverId, "unknown",
                                       _driverConfig.SecondaryDriverId, null);
        }
    }

    private async Task ExecutePhase(PhaseConfig phase)
    {
        Console.WriteLine($"=== Starting phase: {phase.Id} ({phase.Description}) ===");

        int pipelineDepth = phase.PipelineDepth > 0 ? phase.PipelineDepth : DefaultPipelineDepth;

        var clients = await CreateClients(phase).ConfigureAwait(false);
        var commands = CommandFactory.CreateAll(phase.Commands);
        var keyGenerator = KeyGenerator.Create(phase.Keyspace);
        var rateLimiter = phase.HasRpsLimit ? RateLimiter.Create(phase.RpsLimit) : null;
        var metrics = new MetricsCollector();

        await WarmupClients(clients, phase.WarmupRequests).ConfigureAwait(false);

        string status = await ExecuteWorkload(phase, clients, commands, keyGenerator,
                                               rateLimiter, metrics, pipelineDepth).ConfigureAwait(false);

        _metricsWriter.WritePhaseResults(phase.Id, status, phase.Connections, metrics);
        CloseClients(clients);

        Console.WriteLine($"=== Phase {phase.Id} completed: {status} ===");
    }

    private async Task<List<IBenchmarkClient>> CreateClients(PhaseConfig phase)
    {
        Console.WriteLine($"Creating {phase.Connections} connections...");
        var clients = new List<IBenchmarkClient>();
        var cpsLimiter = phase.HasCpsLimit ? RateLimiter.Create(phase.CpsLimit) : null;

        for (int i = 0; i < phase.Connections; i++)
        {
            if (cpsLimiter != null) await cpsLimiter.Acquire().ConfigureAwait(false);
            var client = BenchmarkClientFactory.CreateAndConnect(_host, _port, _driverConfig);
            clients.Add(client);
            if ((i + 1) % 50 == 0)
                Console.WriteLine($"Created {i + 1}/{phase.Connections} connections");
        }

        Console.WriteLine($"All {clients.Count} connections established");
        return clients;
    }

    private async Task WarmupClients(List<IBenchmarkClient> clients, int warmupRequestsPerClient)
    {
        if (warmupRequestsPerClient <= 0) return;
        Console.WriteLine($"Warming up {clients.Count} clients with {warmupRequestsPerClient} PING(s) each...");

        foreach (var client in clients) client.SetWarmupMode(true);

        try
        {
            var tasks = clients.Select(client => Task.Run(async () =>
            {
                for (int r = 0; r < warmupRequestsPerClient; r++)
                    await client.Ping().ConfigureAwait(false);
            })).ToArray();

            await Task.WhenAll(tasks).ConfigureAwait(false);
        }
        finally
        {
            foreach (var client in clients) client.SetWarmupMode(false);
        }

        Console.WriteLine("Warmup completed");
    }

    // Shared mutable state for worker tasks (avoids ref params in async methods)
    private class WorkloadState
    {
        public int Running = 1;
        public long RequestCount;
    }

    private async Task<string> ExecuteWorkload(PhaseConfig phase, List<IBenchmarkClient> clients,
                                                List<ICommand> commands, KeyGenerator keyGenerator,
                                                RateLimiter? rateLimiter, MetricsCollector metrics,
                                                int pipelineDepth)
    {
        var completion = phase.Completion;
        var state = new WorkloadState();
        int workerCount = clients.Count;

        Console.WriteLine($"Spawning {workerCount} worker tasks (pipeline_depth={pipelineDepth})");

        long totalTargetRequests = completion.IsRequestBased ? completion.TotalRequests : -1;
        long endTimeMs = completion.IsDurationBased
            ? DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + (completion.DurationSeconds * 1000)
            : long.MaxValue;

        metrics.Start();
        long startTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

        try
        {
            var workerTasks = new Task[workerCount];

            for (int i = 0; i < workerCount; i++)
            {
                int workerIndex = i;
                var client = clients[i];
                var workerKeyGen = keyGenerator.ForkForThread(workerIndex);
                var workerSelector = new CommandSelector(commands);

                workerTasks[i] = Task.Run(async () =>
                {
                    try
                    {
                        if (pipelineDepth <= 1)
                        {
                            await RunSyncLoop(client, workerSelector, workerKeyGen, rateLimiter,
                                metrics, state, totalTargetRequests, endTimeMs)
                                .ConfigureAwait(false);
                        }
                        else
                        {
                            await RunPipelinedLoop(client, workerSelector, workerKeyGen, rateLimiter,
                                metrics, state, totalTargetRequests, endTimeMs, pipelineDepth)
                                .ConfigureAwait(false);
                        }
                    }
                    catch (Exception)
                    {
                        Interlocked.Exchange(ref state.Running, 0);
                        throw;
                    }
                });
            }

            // Progress logging
            var allDone = Task.WhenAll(workerTasks);
            long lastLogTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            while (!allDone.IsCompleted)
            {
                try { await allDone.WaitAsync(TimeSpan.FromMilliseconds(100)).ConfigureAwait(false); }
                catch (TimeoutException) { }
                catch { break; }

                lastLogTime = LogProgressIfNeeded(Interlocked.Read(ref state.RequestCount), totalTargetRequests,
                                                   lastLogTime, startTime);
            }

            // Check for worker failures
            bool hasFailures = false;
            foreach (var task in workerTasks)
            {
                if (task.IsFaulted)
                {
                    Console.WriteLine($"Worker failed: {task.Exception?.InnerException?.Message}");
                    hasFailures = true;
                }
            }

            if (hasFailures) return "ERROR";
        }
        catch (Exception e)
        {
            Console.WriteLine($"Error during workload execution: {e.Message}");
            return "ERROR";
        }
        finally
        {
            metrics.Stop();
        }

        Console.WriteLine($"All operations completed. Total requests: {Interlocked.Read(ref state.RequestCount)}");
        return "COMPLETED";
    }

    private async Task RunSyncLoop(IBenchmarkClient client, CommandSelector selector,
                                    KeyGenerator keyGen, RateLimiter? rateLimiter,
                                    MetricsCollector metrics, WorkloadState state,
                                    long totalTargetRequests, long endTimeMs)
    {
        while (Volatile.Read(ref state.Running) == 1)
        {
            if (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() >= endTimeMs) break;
            if (totalTargetRequests > 0)
            {
                long claimed = Interlocked.Increment(ref state.RequestCount) - 1;
                if (claimed >= totalTargetRequests)
                {
                    Interlocked.Decrement(ref state.RequestCount);
                    break;
                }
            }

            if (rateLimiter != null) await rateLimiter.Acquire().ConfigureAwait(false);

            var command = selector.Select();
            var result = await command.Execute(client, keyGen).ConfigureAwait(false);
            metrics.Record(result);

            if (totalTargetRequests <= 0)
                Interlocked.Increment(ref state.RequestCount);
        }
    }

    private async Task RunPipelinedLoop(IBenchmarkClient client, CommandSelector selector,
                                         KeyGenerator keyGen, RateLimiter? rateLimiter,
                                         MetricsCollector metrics, WorkloadState state,
                                         long totalTargetRequests, long endTimeMs, int pipelineDepth)
    {
        var pending = new List<Task<CommandResult>>(pipelineDepth);

        // Fill pipeline
        for (int i = 0; i < pipelineDepth && Volatile.Read(ref state.Running) == 1; i++)
        {
            if (!ClaimAndSubmit(client, selector, keyGen, rateLimiter, state,
                               totalTargetRequests, endTimeMs, pending))
                break;
        }

        while (pending.Count > 0 && Volatile.Read(ref state.Running) == 1)
        {
            var completed = await Task.WhenAny(pending).ConfigureAwait(false);
            pending.Remove(completed);

            try
            {
                var result = await completed.ConfigureAwait(false);
                metrics.Record(result);
            }
            catch (Exception e)
            {
                metrics.Record(CommandResult.Failure("UNKNOWN", e.Message));
            }

            if (totalTargetRequests <= 0) Interlocked.Increment(ref state.RequestCount);

            if (Volatile.Read(ref state.Running) == 1)
                ClaimAndSubmit(client, selector, keyGen, rateLimiter, state,
                              totalTargetRequests, endTimeMs, pending);
        }
    }

    private bool ClaimAndSubmit(IBenchmarkClient client, CommandSelector selector,
                                 KeyGenerator keyGen, RateLimiter? rateLimiter,
                                 WorkloadState state, long totalTargetRequests,
                                 long endTimeMs, List<Task<CommandResult>> pending)
    {
        if (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() >= endTimeMs) return false;
        if (totalTargetRequests > 0)
        {
            long claimed = Interlocked.Increment(ref state.RequestCount) - 1;
            if (claimed >= totalTargetRequests)
            {
                Interlocked.Decrement(ref state.RequestCount);
                return false;
            }
        }

        if (rateLimiter != null) rateLimiter.Acquire().GetAwaiter().GetResult();

        var command = selector.Select();
        pending.Add(command.Execute(client, keyGen));
        return true;
    }

    private void CloseClients(List<IBenchmarkClient> clients)
    {
        Console.WriteLine($"Closing {clients.Count} connections...");
        foreach (var client in clients)
        {
            try { client.Dispose(); }
            catch (Exception e) { Console.WriteLine($"Error closing client: {e.Message}"); }
        }
    }

    private long LogProgressIfNeeded(long current, long target, long lastLogTime, long startTime)
    {
        long now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        if (now - lastLogTime < ProgressLogIntervalSeconds * 1000) return lastLogTime;

        long elapsedMs = now - startTime;
        long rate = elapsedMs > 0 ? (current * 1000 / elapsedMs) : 0;

        if (target > 0)
        {
            double percent = current * 100.0 / target;
            Console.WriteLine($"Progress: {current}/{target} requests ({percent:F1} %) - {rate} req/s");
        }
        else
        {
            Console.WriteLine($"Progress: {current} requests - {rate} req/s");
        }

        return now;
    }
}
